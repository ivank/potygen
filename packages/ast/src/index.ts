import { Parser } from '@ikerin/rd-parse';
import { Grammar } from './grammar';
import { QueryTag, WithTag } from './grammar.types';

export {
  AnyCastTag,
  AnyTypeTag,
  ArrayConstructorTag,
  ArrayIndexRangeTag,
  ArrayIndexTag,
  AsTag,
  BetweenTag,
  BinaryExpressionTag,
  BinaryOperatorTag,
  BooleanTag,
  CaseTag,
  CastableDataTypeTag,
  CastTag,
  CollateTag,
  ColumnsTag,
  ColumnTag,
  CombinationTag,
  ComparationExpressionTag,
  ComparationOperatorTag,
  ConflictConstraintTag,
  ConflictTag,
  ConflictTargetTag,
  ConstantTag,
  CountTag,
  CTETag,
  DataTypeTag,
  DefaultTag,
  DeleteTag,
  DistinctTag,
  DoNothingTag,
  DoUpdateTag,
  ElseTag,
  ExpressionListTag,
  ExpressionTag,
  FromListItemTag,
  FromListTag,
  FromTag,
  FunctionTag,
  GroupByTag,
  HavingTag,
  IdentifierTag,
  InsertTag,
  JoinOnTag,
  JoinTag,
  JoinTypeTag,
  JoinUsingTag,
  LimitTag,
  NullTag,
  NumberTag,
  OffsetTag,
  OperatorExpressionTag,
  OrderByItemTag,
  OrderByTag,
  OrderDirectionTag,
  ParameterTag,
  PgCastTag,
  QuotedNameTag,
  ReturningListItemTag,
  ReturningTag,
  RowTag,
  SelectListItemTag,
  SelectListTag,
  SelectParts,
  SelectTag,
  SetItemTag,
  SetListTag,
  SetMapTag,
  SetTag,
  SqlTag,
  StarIdentifierTag,
  StarTag,
  StringTag,
  TableTag,
  Tag,
  TypeArrayTag,
  TypeTag,
  UnaryExpressionTag,
  UnaryOperatorTag,
  UpdateFromTag,
  UpdateTag,
  UsingTag,
  ValuesListTag,
  ValuesTag,
  WhenTag,
  WhereTag,
  WithTag,
} from './grammar.types';

export {
  isAnyCast,
  isAnyType,
  isArrayConstructor,
  isArrayIndex,
  isArrayIndexRange,
  isAs,
  isBetween,
  isBinaryExpression,
  isBinaryOperator,
  isBoolean,
  isCase,
  isCast,
  isCastableDataType,
  isCollateTag,
  isColumn,
  isColumns,
  isCombination,
  isComparationExpression,
  isComparationOperator,
  isConflict,
  isConflictConstraint,
  isConflictTarget,
  isConstant,
  isCount,
  isCTETag,
  isDataType,
  isDefault,
  isDelete,
  isDistinct,
  isDoNothing,
  isDoUpdate,
  isElse,
  isExpression,
  isExpressionList,
  isFrom,
  isFromList,
  isFromListItem,
  isFunction,
  isGroupBy,
  isHaving,
  isIdentifier,
  isInsertTag,
  isJoin,
  isJoinOn,
  isJoinType,
  isJoinUsing,
  isLimit,
  isNull,
  isNumber,
  isOffset,
  isOperatorExpression,
  isOrderBy,
  isOrderByItem,
  isOrderDirection,
  isParameter,
  isPgCast,
  isQuotedName,
  isReturning,
  isReturningListItem,
  isRow,
  isSelect,
  isSelectList,
  isSelectListItem,
  isSelectTag,
  isSet,
  isSetItem,
  isSetList,
  isSetMap,
  isStar,
  isStarIdentifier,
  isString,
  isTable,
  isType,
  isTypeArray,
  isUnaryExpression,
  isUnaryOperator,
  isUpdate,
  isUpdateFrom,
  isUsing,
  isValues,
  isValuesList,
  isWhen,
  isWhere,
  isWithTag,
  isWrappedExpression,
} from './grammar.guards';

export {
  chunk,
  first,
  groupBy,
  identity,
  initial,
  isDiffBy,
  isEmpty,
  isEqual,
  isNil,
  isObject,
  isUnique,
  isUniqueBy,
  last,
  length,
  orderBy,
  tail,
  toMilliseconds,
} from './util';

export { Grammar } from './grammar';

export const parser = Parser<QueryTag | WithTag>(Grammar);
