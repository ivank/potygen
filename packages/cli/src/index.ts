export {
  DataTable,
  DataFunction,
  DataEnum,
  DataView,
  DataComposite,
  LoadedDataTable,
  LoadedDataComposite,
  DataViewRaw,
  DataViewParsed,
  LoadedDataView,
  LoadedDataFunction,
  LoadedDataEnum,
  Data,
  LoadedData,
  LoadedParam,
  LoadedResult,
  LoadedQueryInterface,
  LoadedFunction,
  LoadedComposite,
  LoadedSourceTable,
  LoadedSourceQuery,
  LoadedSourceView,
  LoadedSourceValues,
  LoadedSource,
  LoadedContext,
  TemplateTagQuery,
  ParsedTypescriptFile,
  ParsedSqlFile,
  ParsedFile,
  LoadedTemplateTagQuery,
  LoadedTypescriptFile,
  LoadedSqlFile,
  LoadedFile,
  Logger,
  LoadContext,
  QualifiedName,
} from './types';

export {
  isDataTable,
  isDataFunction,
  isDataEnum,
  isLoadedDataTable,
  isLoadedDataFunction,
  isLoadedDataEnum,
  isLoadedDataView,
  isLoadedDataComposite,
} from './guards';

export { toLoadedContext, loadQueryInterfacesData, toLoadedQueryInterface, loadData, extractDataSources } from './load';

export { Config, ConfigType, FullConfigType, toConfig } from './config';

export { emitLoadedFile, toTypeSource } from './emit';
export { glob } from './glob';
