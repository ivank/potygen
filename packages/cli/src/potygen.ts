import { Logger } from '@potygen/potygen';
import { existsSync, readFileSync } from 'fs';
import { inspect, promisify } from 'util';
import { Client } from 'pg';
import { pipeline } from 'stream';
import { Command, createCommand } from 'commander';
import { QueryLoader, SqlRead } from './traverse';
import { ConfigType, toConfig } from './config';

const asyncPipeline = promisify(pipeline);

enum LogLevel {
  error,
  info,
  debug,
}

class LogLevelConsole implements Logger {
  constructor(public level: LogLevel) {}

  error(...args: any[]): void {
    if (this.level >= LogLevel.error) {
      console.info(...args);
    }
  }

  info(...args: any[]): void {
    if (this.level >= LogLevel.info) {
      console.info(...args);
    }
  }

  debug(...args: any[]): void {
    if (this.level >= LogLevel.debug) {
      console.debug(...args);
    }
  }
}

export const potygen = (overwriteLogger?: Logger): Command =>
  createCommand('potygen')
    .description('Convert postgres query files inty typescript types')
    .version('0.5.11')
    .option('-c, --config <config>', 'A configuration file to load', 'potygen.config.json')
    .option('-f, --files <files>', 'A glob pattern to search files by (default: "**/*.sql")')
    .option('-w, --watch', 'Watch for file changes and update live')
    .option('-v, --verbose', 'Show verbose logs')
    .option('-s, --silent', 'Only show error logs')
    .option('-p, --typePrefix <typePrefix>', 'Prefix generated types')
    .option('-r, --root <root>', `Set the root directory (default: ${process.cwd()})`)
    .option(
      '-n, --connection <connection>',
      'Connection to the postgres database. URI (default: "postgres://localhost:5432/db")',
    )
    .option(
      '-t, --template <template>',
      'A template of the path, where to generate the typescript type files. The parameters are the response from node\'s path.parse function (default: "{{dir}}/{{name}}.queries.ts")',
    )
    .action(async (options: ConfigType & { config: string }) => {
      const { config, ...rest } = options;
      const { root, connection, watch, files, template, verbose, silent, typePrefix } = toConfig({
        ...(existsSync(config) ? JSON.parse(readFileSync(config, 'utf-8')) : {}),
        ...rest,
      });

      const logger =
        overwriteLogger ?? new LogLevelConsole(verbose ? LogLevel.debug : silent ? LogLevel.error : LogLevel.info);

      logger.debug(`Potygen Config: ${inspect({ root, connection, watch, files, template })}`);

      const db = new Client(connection);
      await db.connect();
      try {
        const sqls = new SqlRead({ path: files, root, watch, logger });
        const sink = new QueryLoader({ db, root, template, logger, typePrefix });

        logger.info(`Potygen started processing ("${files}", watch: ${watch})`);

        await asyncPipeline(sqls, sink);
      } catch (error) {
        if (process.env.POTYGEN_DEBUG && error instanceof Error) {
          logger.error(error.stack);
        } else {
          logger.error(String(error));
        }
      } finally {
        await db.end();
      }
    });
