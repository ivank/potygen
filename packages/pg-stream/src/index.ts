export { toInternalCursorGenerator } from './internal-cursor-generator';
export { ToEachBatch, ToEachBatchOptions, toEachBatch } from './each-batch';
export { ToReadable, ToReadableOptions, toReadable } from './readable';
export {
  ToAsyncIterator,
  toAsyncIterator,
  ToIteratorOptions,
  ToAsyncBatchIterator,
  toAsyncBatchIterator,
} from './iterator';
