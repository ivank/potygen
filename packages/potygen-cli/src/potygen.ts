import { Command, createCommand } from 'commander';
import { Logger } from './types';
import { QueryLoader, SqlRead } from './traverse';
import { Client } from 'pg';
import { promisify } from 'util';
import { pipeline } from 'stream';
import { existsSync, readFileSync } from 'fs';
import { Record, String, Optional, Boolean, Static } from 'runtypes';

const Config = Record({
  files: Optional(String),
  watch: Optional(Boolean),
  root: Optional(String),
  connection: Optional(String),
  template: Optional(String),
});

type ConfigType = Static<typeof Config>;

const asyncPipeline = promisify(pipeline);

export const potygen = (logger: Logger = console): Command =>
  createCommand('potygen')
    .description('Convert postgres query files inty typescript types')
    .option('-c, --config <config>', 'A configuration file to load', 'potygen.config.json')
    .option('-f, --files <files>', 'A glob pattern to search files by (default: "**/*.sql")')
    .option('-w, --watch', 'Watch for file changes and update live')
    .option('-r, --root <root>', `Set the root directory (default: ${process.cwd()})`)
    .option(
      '-n, --connection <connection>',
      'Connection to the postgres database. URI (default: "postgres://localhost:5432/db")',
    )
    .option(
      '-t, --template <template>',
      'A template of the path, where to generate the typescript type files. The parameters are the response from node\'s path.parse function (default: "{{dir}}/{{name}}.queries.ts")',
    )
    .action(async (options: ConfigType & { config: string }) => {
      const config: ConfigType = existsSync(options.config)
        ? Config.check(JSON.parse(readFileSync(options.config, 'utf-8')))
        : {};

      const { root, connection, watch, files, template } = {
        files: '**/*.sql',
        root: process.cwd(),
        template: '{{dir}}/{{name}}.queries.ts',
        connection: 'postgres://localhost:5432/db',
        watch: false,
        ...config,
        ...options,
      };

      if (process.env.POTYGEN_DEBUG) {
        console.info('Potygen Config', { root, connection, watch, files, template });
      }

      const db = new Client(connection);
      await db.connect();
      try {
        const sqls = new SqlRead({ path: files, root, watch, logger });
        const sink = new QueryLoader({ db, root, template });

        logger.info(`Potygen started processing ("${files}", watch: ${watch})`);

        await asyncPipeline(sqls, sink);
      } catch (error) {
        if (process.env.POTYGEN_DEBUG && error instanceof Error) {
          logger.error(error.stack);
        } else {
          logger.error(global.String(error));
        }
      } finally {
        await db.end();
      }
    });
