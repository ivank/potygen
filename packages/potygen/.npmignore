.*
test/
src/
examples/
coverage/
tsconfig.json
