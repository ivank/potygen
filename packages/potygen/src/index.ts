export {
  AstTag,
  AnyCastTag,
  AnyTypeTag,
  ArrayColumnIndexTag,
  ArrayConstructorTag,
  ArrayIndexRangeTag,
  ArrayIndexTag,
  BeginTag,
  BinaryExpressionTag,
  BinaryOperatorTag,
  BitStringTag,
  BooleanTag,
  CaseSimpleTag,
  CaseTag,
  CastableDataTypeTag,
  CastTag,
  CollateTag,
  ColumnsTag,
  ColumnTag,
  CombinationTag,
  CommentTag,
  ComparationArrayInclusionTypeTag,
  ComparationArrayOperatorTag,
  ComparationArrayTypeTag,
  ComparationArrayInclusionTag,
  ComparationArrayTag,
  ExistsTag,
  CompositeAccessTag,
  ConflictConstraintTag,
  ConflictTag,
  ConflictTargetTag,
  ConstantTag,
  CountTag,
  CTENameTag,
  CTETag,
  CTEValuesListTag,
  CTEValuesTag,
  CustomQuotedStringTag,
  DataTypeTag,
  DefaultTag,
  DeleteTag,
  DimensionTag,
  DistinctTag,
  DollarQuotedStringTag,
  DoNothingTag,
  DoUpdateTag,
  ElseTag,
  EmptyLeafSqlTag,
  EmptyLeafTag,
  EscapeStringTag,
  ExpressionListTag,
  ExpressionTag,
  ExtractFieldTag,
  ExtractTag,
  FilterTag,
  FromListItemTag,
  FromListTag,
  FromTag,
  FunctionArgTag,
  FunctionTag,
  GroupByTag,
  HavingTag,
  HexademicalStringTag,
  IdentifierTag,
  InsertTag,
  JoinOnTag,
  JoinTag,
  JoinTypeTag,
  JoinUsingTag,
  LeafSqlTag,
  LimitAllTag,
  LimitTag,
  NamedSelectTag,
  NodeSqlTag,
  NodeTag,
  NullTag,
  NumberTag,
  OffsetTag,
  OperatorExpressionTag,
  OrderByItemTag,
  OrderByTag,
  OrderDirectionTag,
  ParameterTag,
  ParameterPickTag,
  PgCastTag,
  QualifiedIdentifierTag,
  QueryTag,
  QuotedIdentifierTag,
  ReturningListItemTag,
  ReturningTag,
  RollbackTag,
  RowTag,
  RowKeywardTag,
  SavepointTag,
  SelectListItemTag,
  SelectListTag,
  SelectParts,
  SelectTag,
  SetItemTag,
  SetArrayItemTag,
  SetListTag,
  SetMapTag,
  SetTag,
  SqlTag,
  StarIdentifierTag,
  StarTag,
  StringTag,
  TableTag,
  TableWithJoinTag,
  Tag,
  TernaryExpressionTag,
  TernaryOperatorTag,
  TernarySeparatorTag,
  ArrayTypeTag,
  TypeTag,
  UnaryExpressionTag,
  UnaryOperatorTag,
  UnquotedIdentifierTag,
  UpdateFromTag,
  UpdateTag,
  UsingTag,
  ValuesListTag,
  ValuesTag,
  WhenTag,
  WhereTag,
  WithTag,
  WrappedExpressionTag,
  SqlName,
} from './grammar.types';

export {
  isAnyCast,
  isAnyType,
  isArrayColumnIndex,
  isArrayConstructor,
  isArrayIndex,
  isArrayIndexRange,
  isAs,
  isBinaryExpression,
  isBinaryOperator,
  isBitString,
  isBoolean,
  isCase,
  isCaseSimple,
  isCast,
  isCastableDataType,
  isCollate,
  isColumn,
  isColumns,
  isCombination,
  isComment,
  isComparationArrayInclusionType,
  isComparationArrayOperator,
  isComparationArrayType,
  isComparationArrayInclusion,
  isComparationArray,
  isExists,
  isCompositeAccess,
  isConflict,
  isConflictConstraint,
  isConflictTarget,
  isConflictTargetIndex,
  isConstant,
  isCount,
  isCTE,
  isCTEName,
  isCTEValues,
  isCTEValuesList,
  isCustomQuotedString,
  isDataType,
  isDefault,
  isDelete,
  isDistinct,
  isDollarQuotedString,
  isDoNothing,
  isDoUpdate,
  isElse,
  isEmptyLeaf,
  isEscapeString,
  isExpression,
  isExpressionList,
  isFilter,
  isFrom,
  isFromList,
  isFromListItem,
  isFunction,
  isFunctionArg,
  isGroupBy,
  isHaving,
  isHexademicalString,
  isIdentifier,
  isInsert,
  isInteger,
  isJoin,
  isJoinOn,
  isJoinType,
  isJoinUsing,
  isLeaf,
  isLimit,
  isLimitAll,
  isNamedSelect,
  isNode,
  isNull,
  isNumber,
  isOffset,
  isOperatorExpression,
  isOrderBy,
  isOrderByItem,
  isOrderDirection,
  isParameter,
  isParameterPick,
  isPgCast,
  isQualifiedIdentifier,
  isQuotedIdentifier,
  isReturning,
  isReturningListItem,
  isRow,
  isRowKeyward,
  isSelect,
  isSelectList,
  isSelectListItem,
  isSet,
  isSetItem,
  isSetArrayItem,
  isSetList,
  isSetMap,
  isStar,
  isStarIdentifier,
  isString,
  isTable,
  isTableWithJoin,
  isTernaryExpression,
  isArrayType,
  isUnaryExpression,
  isUnaryOperator,
  isUnquotedIdentifier,
  isUpdate,
  isUpdateFrom,
  isUsing,
  isValues,
  isValuesList,
  isWhen,
  isWhere,
  isWith,
  isWrappedExpression,
} from './grammar.guards';

export {
  chunk,
  first,
  groupBy,
  identity,
  initial,
  isDiffBy,
  isEmpty,
  isEqual,
  isNil,
  isObject,
  isUnique,
  isUniqueBy,
  last,
  orderBy,
  tail,
  toMilliseconds,
  range,
  findLastIndex,
} from './util';

export { parser, partialParser } from './grammar';

export {
  TypeString,
  TypeNumber,
  TypeBoolean,
  TypeDate,
  TypeNull,
  TypeJson,
  TypeUnknown,
  TypeLoadCoalesce,
  TypeLoadRecord,
  TypeLoadFunction,
  TypeLoadColumn,
  TypeLoadStar,
  TypeLoadFunctionArgument,
  TypeLoadOperator,
  TypeLoadNamed,
  TypeLoadArray,
  TypeLoadAsArray,
  TypeAny,
  TypeLoadArrayItem,
  TypeContext,
  TypeLiteral,
  BaseTypeLoad,
  BaseType,
  TypeNullable,
  TypeLoadUnion,
  TypeBuffer,
  TypeUnion,
  TypeArray,
  Type,
  TypeOrLoad,
  Result,
  Param,
  ParamPick,
  Source,
  SourceTable,
  SourceQuery,
  QueryInterface,
  TypeLoadObjectLiteral,
  TypeObjectLiteral,
  TypeLoadCompositeAccess,
  TypeComposite,
  TypeLoadOptional,
  TypeOptional,
  TypeLoadColumnCast,
  OperatorVariantPart,
  OperatorVariant,
  TypeName,
} from './query-interface.types';

export {
  isType,
  isTypeString,
  isTypeNumber,
  isTypeBoolean,
  isTypeDate,
  isTypeNull,
  isTypeJson,
  isTypeUnknown,
  isTypeLoadRecord,
  isTypeLoadFunction,
  isTypeLoadColumn,
  isTypeLoadStar,
  isTypeLoadFunctionArgument,
  isTypeLoadOperator,
  isTypeLoadNamed,
  isTypeCoalesce,
  isTypeLoadArray,
  isTypeLoadUnion,
  isTypeArray,
  isTypeNullable,
  isTypeLiteral,
  isTypeUnion,
  isTypeLoadObjectLiteral,
  isTypeAny,
  isTypeLoadArrayItem,
  isTypeLoadAsArray,
  isTypeObjectLiteral,
  isTypeComposite,
  isTypeLoadCompositeAccess,
  isTypeLoadOptional,
  isTypeOptional,
  isTypeLoadColumnCast,
  isSourceTable,
  isSourceQuery,
  isSourceValues,
  isTypeEqual,
  isTypeLoad,
} from './query-interface.guards';

export { SqlInterface, QueryConfig, QuerySource, SqlResult, SqlDatabase, Query, MapQuery, Json } from './sql.types';
export { oneResult, maybeOneResult, mapResult } from './sql.helpers';
export {
  toParams,
  toQueryInterface,
  toConstantBinaryOperatorVariant,
  toAliasedPgType,
  toPgTypeConstant,
  toSources,
  toQueryFrom,
  toQueryResults,
} from './query-interface';
export { sql, toQuery, toQueryConfig } from './sql';
export {
  DatabaseError,
  isDatabaseError,
  PotygenError,
  PotygenDatabaseError,
  PotygenNotFoundError,
  LoadError,
  ParseError,
  ParsedTypescriptFileLoadError,
  ParsedSqlFileLoadError,
} from './errors';

export {
  DataTable,
  DataFunction,
  DataEnum,
  DataView,
  DataComposite,
  LoadedDataTable,
  LoadedDataComposite,
  DataViewRaw,
  DataViewParsed,
  LoadedDataView,
  LoadedDataFunction,
  LoadedDataEnum,
  Data,
  LoadedData,
  LoadedParam,
  LoadedResult,
  LoadedQueryInterface,
  LoadedFunction,
  LoadedComposite,
  LoadedSourceTable,
  LoadedSourceQuery,
  LoadedSourceView,
  LoadedSourceValues,
  LoadedSource,
  LoadedContext,
  TemplateTagQuery,
  ParsedTypescriptFile,
  ParsedSqlFile,
  ParsedFile,
  LoadedTemplateTagQuery,
  LoadedTypescriptFile,
  LoadedSqlFile,
  LoadedFile,
  Logger,
  LoadContext,
  QualifiedName,
  Path,
  PathItem,
  CompletionEntry,
  InfoContext,
  InfoLoadedQuery,
  QuickInfo,
  Cache,
} from './load.types';

export {
  isDataTable,
  isDataFunction,
  isDataEnum,
  isLoadedDataTable,
  isLoadedDataFunction,
  isLoadedDataEnum,
  isLoadedDataView,
  isLoadedDataComposite,
  isLoadedSourceView,
  isLoadedSourceTable,
  isLoadedSourceValues,
  isLoadedSource,
  isLoadedSourceUnknown,
} from './load.guards';

export {
  toLoadedContext,
  loadQueryInterfacesData,
  toLoadedQueryInterface,
  loadData,
  loadAllData,
  extractDataSources,
  filterUnknownLoadedContext,
  throwOnUnknownLoadedContext,
} from './load';

export { LRUCache } from './inspect/cache';
export { closestParent, closestParentPath, toPath } from './inspect/path';
export { completionAtOffset, toInfoContext, quickInfoAtOffset, inspectError } from './inspect/inspect';
