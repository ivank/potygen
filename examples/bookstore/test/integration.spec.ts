import { createApp } from '../src/app';
import * as pgp from 'pg-promise';
import { createServer } from 'http';
import axios from 'axios';
import { sql } from '@potygen/potygen';

const connection = process.env.POSTGRES_CONNECTION ?? 'postgres://potygen:dev-pass@localhost:5432/potygen';

describe('Integration test', () => {
  it('Test requests', async () => {
    const db = pgp()(connection);

    await sql`DELETE FROM books`(db, {});

    await sql`
      INSERT INTO "books" (
        id,
        isbn,
        title,
        author_id
      )
      VALUES
        (
          1,
          '0-333-45430-8',
          'Consider Phlebas',
          1
        ),
        (
          2,
          '0-333-47110-5',
          'The Player of Games',
          1
        ),
        (
          3,
          '0-312-85182-0',
          'A Fire Upon the Deep',
          2
        )
      `(db, {});

    const app = createApp(db);

    const server = createServer(app);
    await new Promise((resolve, reject) =>
      server.listen(3201, (error?: Error) => (error ? reject(error) : resolve(undefined))),
    );

    const { data } = await axios.get('http://localhost:3201/books/1');

    expect(data).toMatchSnapshot('Books');

    await axios.post('http://localhost:3201/books', { isbn: '11-22-33', title: 'Test', authorId: 1 });

    const { data: modifiedData } = await axios.get('http://localhost:3201/books/1');

    expect(modifiedData).toHaveLength(3);

    await db.$pool.end();
    await new Promise((resolve, reject) => server.close((error) => (error ? reject(error) : resolve(undefined))));
  });
});
