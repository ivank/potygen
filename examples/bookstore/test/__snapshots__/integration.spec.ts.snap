// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration test Test requests: Books 1`] = `
Array [
  Object {
    "authorName": "Iain Banks",
    "copiesSold": "0",
    "id": 1,
    "isbn": "0-333-45430-8",
    "title": "Consider Phlebas",
  },
  Object {
    "authorName": "Iain Banks",
    "copiesSold": "0",
    "id": 2,
    "isbn": "0-333-47110-5",
    "title": "The Player of Games",
  },
]
`;
